content/userdefinedfunctions.md

```diff
@@ -0,0 +1,189 @@
+---
+id: userdefinedfunctions
+title: User Defined Functions
+sidebar_label: User Defined Functions
+slug: /
+---
+
+User Defined functions provide a lightweight way to implement custom behaviour outside of what Hypi offers. The functions are executed on the same server where the query gets processed.
+
+Currently, you may use two programming language for the user defined functions.
+
++ Groovy: The entire Groovy syntax is available on the hypi platform. Groovy is general a general programming language. Adding custom behaviour using Groovy becomes quick, easy and performant.
+
++ Velocity: The velocity template language is in place to output text. This text may execute dynamic behaviour. E.g., customizing the body of an email, substituting receiver’s name from a variable etc.
+
+> Let’s check how to implement user defined functions. 
+
+We will start with a simple schema as follows: 
+```java
+type Query {
+    inlineGroovyFunction(a: String, b: Int, c: Boolean):Json @tan(type:Groovy, inline: """
+        def map = new java.util.LinkedHashMap()
+        map.put("a", a)
+        map.put("b", b)
+        map.put("c", c)
+        return map
+    """)
+    inlineVelocityFunction(a: String, b: Int, c: Boolean):
+    String @tan(type:Velocity, inline: "$a,$b,$c")
+}
+```
+
+The schema has two user defined functions. `inlineGroovyFunction` and `inlineVelocityFunction` You may define the user defined functions either inline or through a script. `@tan` directive is used to add the functions. 
+
+`inlineGroovyFunction` function puts value of variables a,b and c in a LinkedHashMap and the map is returned in the form of JSON data. `inlineVelocityFunction` uses velocity template to return the formatted string output of variables a,b, and c. You may execute the user defined function as any other GraphQL query.
+
+```java
+{
+  inlineGroovyFunction(a:"val1", b:23, c: true)
+}
+OR
+{
+  inlineVelocityFunction(a:"val1", b:23, c: true)
+}
+
+```
+Result:
+```json
+{
+  "data": {
+    "inlineGroovyFunction": {
+      "a": "val1",
+      "b": 23,
+      "c": true
+    }
+  }
+}
+OR
+{
+  "data": {
+    "inlineVelocityFunction": "val1,23,true"
+  }
+}
+```
+You may pass the variable parameters to the function as well. Just like this:
+```java
+query inlineVelocityFunction($a: String, $b: Int, $c: Boolean){
+  inlineVelocityFunction(a: $a, b: $b, c: $c)
+}
+```
+As you can see in `inlineGroovyFunction`, Java classes are also available to use inline. You may also use a simple Groovy like version as follows.
+```java
+inlineGroovyFunction(a: String, b: Int, c: Boolean):Json @tan(type:Groovy, inline: """
```

Comment:
Does this work? Don't you need to quote the keys? I can't remember
```java
inlineGroovyFunction(a: String, b: Int, c: Boolean):Json @tan(type:Groovy, inline: "return ['a':a, 'b':b, 'c':c]")
```

Reply:
yes..it works